src/Pate/Monad.hs

```diff
@@ -163,12 +175,29 @@ data EquivEnv sym arch where
     , envTocs :: PA.HasTOCReg arch => (TOC.TOC (MM.ArchAddrWidth arch), TOC.TOC (MM.ArchAddrWidth arch))
     -- ^ table of contents for the original and patched binaries, if defined by the
     -- architecture
-    , envCurrentFunc :: PatchPair arch
+    , envCurrentFunc :: BlockPair arch
     -- ^ start of the function currently under analysis
     , envCurrentFrame :: AssumptionFrame sym
     -- ^ the current assumption frame, accumulated as assumptions are added
+    , envNonceGenerator :: N.NonceGenerator IO (PF.ProofScope (PFI.ProofSym sym arch))
+    , envParentNonce :: Some (PF.ProofNonce (PFI.ProofSym sym arch))
+    -- ^ nonce of the parent proof node currently in scope
     } -> EquivEnv sym arch
 
+freshNonce :: EquivM sym arch (N.Nonce (PF.ProofScope (PFI.ProofSym sym arch)) tp)
+freshNonce = do
+  gen <- asks envNonceGenerator
+  liftIO $ N.freshNonce gen
+
+withProofNonce ::
+  forall tp sym arch a.
+  (PF.ProofNonce (PFI.ProofSym sym arch) tp -> EquivM sym arch a) ->
```

Comment:
It's not, because it's only allocating a single proof nonce representing the "current" node (setting it to the envParentNonce field within the continuation), which is a known type.